### PR TITLE
YJIT: 32 bit comparison on shape id

### PR DESCRIPTION
**Pull request by tenderlove**
This commit changes the shape id comparisons to use a 32 bit comparison rather than 64 bit.  That means we don't need to load the shape id to a register on x86 machines.

I've also added a guard for "Object" types, and added a way to upgrade the type to an object on the context.

Unfortunately this patch also generates an extra guard for the object type, but that guard can be amortized over many ivar reads.  Once we incorporate T_OBJECT / T_* information in to the shape, we can remove this guard.

Given the following program:

```ruby
class Foo
  def initialize
    @foo = 1
    @bar = 1
  end

  def read
    [@foo, @bar]
  end
end

foo = Foo.new
foo.read
foo.read
foo.read
foo.read
foo.read

puts RubyVM::YJIT.disasm(Foo.instance_method(:read))
```

The machine code we generated _before_ this change is like this:

```
== BLOCK 1/4, ISEQ RANGE [0,3), 65 bytes ======================
  # getinstancevariable
  0x559a18623023: mov rax, qword ptr [r13 + 0x18]
  # guard object is heap
  0x559a18623027: test al, 7
  0x559a1862302a: jne 0x559a1862502d
  0x559a18623030: cmp rax, 4
  0x559a18623034: jbe 0x559a1862502d
  # guard shape, embedded, and T_OBJECT
  0x559a1862303a: mov rcx, qword ptr [rax]
  0x559a1862303d: movabs r11, 0xffff00000000201f
  0x559a18623047: and rcx, r11
  0x559a1862304a: movabs r11, 0xb000000002001
  0x559a18623054: cmp rcx, r11
  0x559a18623057: jne 0x559a18625046
  0x559a1862305d: mov rax, qword ptr [rax + 0x18]
  0x559a18623061: mov qword ptr [rbx], rax

== BLOCK 2/4, ISEQ RANGE [3,6), 0 bytes =======================
== BLOCK 3/4, ISEQ RANGE [3,6), 47 bytes ======================
  # gen_direct_jmp: fallthrough
  # getinstancevariable
  # regenerate_branch
  # getinstancevariable
  # regenerate_branch
  0x559a18623064: mov rax, qword ptr [r13 + 0x18]
  # guard shape, embedded, and T_OBJECT
  0x559a18623068: mov rcx, qword ptr [rax]
  0x559a1862306b: movabs r11, 0xffff00000000201f
  0x559a18623075: and rcx, r11
  0x559a18623078: movabs r11, 0xb000000002001
  0x559a18623082: cmp rcx, r11
  0x559a18623085: jne 0x559a18625099
  0x559a1862308b: mov rax, qword ptr [rax + 0x20]
  0x559a1862308f: mov qword ptr [rbx + 8], rax
```

After this change, it's like this:

```
== BLOCK 1/4, ISEQ RANGE [0,3), 58 bytes ======================
  # getinstancevariable
  0x5608ccada023: mov rax, qword ptr [r13 + 0x18]
  # guard object is heap
  0x5608ccada027: test al, 7
  0x5608ccada02a: jne 0x5608ccadc02d
  0x5608ccada030: cmp rax, 4
  0x5608ccada034: jbe 0x5608ccadc02d
  # guard object is object
  0x5608ccada03a: mov rcx, qword ptr [rax]
  0x5608ccada03d: and rcx, 0x1f
  0x5608ccada041: cmp rcx, 1
  0x5608ccada045: jne 0x5608ccadc02d
  # guard shape
  0x5608ccada04b: cmp word ptr [rax + 6], 0x13
  0x5608ccada050: jne 0x5608ccadc046
  0x5608ccada056: mov rax, qword ptr [rax + 0x10]
  0x5608ccada05a: mov qword ptr [rbx], rax

== BLOCK 2/4, ISEQ RANGE [3,6), 0 bytes =======================
== BLOCK 3/4, ISEQ RANGE [3,6), 23 bytes ======================
  # gen_direct_jmp: fallthrough
  # getinstancevariable
  # regenerate_branch
  # getinstancevariable
  # regenerate_branch
  0x5608ccada05d: mov rax, qword ptr [r13 + 0x18]
  # guard shape
  0x5608ccada061: cmp word ptr [rax + 6], 0x13
  0x5608ccada066: jne 0x5608ccadc099
  0x5608ccada06c: mov rax, qword ptr [rax + 0x18]
  0x5608ccada070: mov qword ptr [rbx + 8], rax
```

The first ivar read is a bit more complex, but the second ivar read is much simpler.  I think eventually we could teach the context about the shape, then emit only one shape guard.

### CODE REVIEW

File: yjit/src/codegen.rs
Comment (tenderlove) [reply yjit/src/codegen.rs]: Good catch.  I've fixed this.  Thank you!

**Comment by tenderlove** [comment]
I ran YJIT Bench with this branch:

```
before: ruby 3.2.0dev (2022-11-16T18:09:15Z master 3eb7a6521c) +YJIT [x86_64-linux]
after: ruby 3.2.0dev (2022-11-15T21:31:44Z 32-bit-compare 962dc0eafb) +YJIT [x86_64-linux]

-------------  -----------  ----------  ----------  ----------  ------------  -------------
bench          before (ms)  stddev (%)  after (ms)  stddev (%)  before/after  after 1st itr
30k_ifelse     222.8        0.0         230.2       0.1         0.97          0.98         
30k_methods    536.6        1.0         559.7       0.0         0.96          0.96         
activerecord   69.3         0.2         69.0        0.2         1.00          1.01         
binarytrees    139.0        2.6         138.9       0.3         1.00          0.99         
cfunc_itself   30.7         0.1         31.4        0.5         0.98          0.97         
chunky_png     443.0        0.4         481.9       0.2         0.92          0.92         
erubi          199.6        0.9         203.3       0.9         0.98          0.98         
erubi_rails    12.4         2.8         12.3        2.8         1.01          0.99         
etanni         339.5        0.4         345.3       0.4         0.98          0.99         
fannkuchredux  1446.7       1.7         1439.4      0.2         1.01          1.00         
fib            41.7         0.2         41.7        0.1         1.00          0.99         
getivar        23.3         0.1         24.3        0.1         0.96          1.01         
hexapdf        1506.0       2.5         1521.7      2.4         0.99          1.01         
keyword_args   39.0         0.3         39.1        0.2         1.00          1.00         
lee            654.4        1.5         651.5       1.3         1.00          1.01         
liquid-render  77.9         1.6         77.1        1.6         1.01          0.99         
mail           106.0        0.2         106.2       0.2         1.00          0.99         
nbody          63.1         0.2         60.3        0.4         1.05          1.05         
optcarrot      2150.5       0.6         2117.3      0.6         1.02          1.01         
psych-load     1355.3       0.1         1351.2      0.1         1.00          1.00         
railsbench     1297.7       1.4         1277.4      1.9         1.02          1.01         
respond_to     18.1         3.5         20.8        0.2         0.87          0.86         
ruby-lsp       51.1         20.5        51.1        20.9        1.00          0.99         
rubykon        4849.6       0.4         4830.2      0.4         1.00          1.00         
setivar        30.3         0.3         31.2        8.1         0.97          1.01         
str_concat     26.5         2.1         26.5        2.3         1.00          0.99         
-------------  -----------  ----------  ----------  ----------  ------------  -------------
Legend:
- before/after: ratio of before/after time. Higher is better for after. Above 1 represents a speedup.
- after 1st itr: ratio of before/after time for the first benchmarking iteration.
```

Looks like `nbody` is faster.  `respond_to` is slower, but I don't understand why.  I looked at the benchmark and it has nothing to do with instance variables, so I don't understand.

File: yjit/src/backend/arm64/mod.rs
Comment (maximecb) [inline yjit/src/backend/arm64/mod.rs:353]: What does this change do?

File: shape.c
Comment (maximecb) [reply shape.c:329]: Also may want to give it a more descriptive name like `SHAPE_ID_NUM_BITS` ?

File: yjit/src/backend/arm64/mod.rs
Comment (tenderlove) [reply yjit/src/backend/arm64/mod.rs:353]: In the case an immediate cannot be used in one instruction (for example `4097`) the backend will move it in to a register.  The problem is that it always moves it to a 64 bit register when we may have requested a comparison with a 32 bit register.  If the number of bits in the split register doesn't match the number of bits in the comparison register, then we convert the split register to match the other register.

I'm not 100% sure if this is the right solution, but the test below exercises this case, and we ran in to this problem when trying to do the 16 / 32 bit comparison for the shape id.

**Comment by tenderlove** [comment]
> I think it's good that the shape id comparison can be made smaller, but the perf drop on `respond_to` is pretty big. It might be worth looking at the disasm. It's possible that the changes to the splitting code may have caused the difference.

I'll take a look at the disasm.  I don't think the splitting impacted it though as I've only messed with splitting on arm64 and the benchmarks were on x86.

File: shape.c
Comment (maximecb) [reply shape.c:329]: Makes sense. I do think the name could be better (aren't bits always one bit in size? 😅)

File: yjit/src/backend/arm64/mod.rs
Comment (maximecb) [reply yjit/src/backend/arm64/mod.rs:353]: Makes sense 👍

File: yjit/src/codegen.rs
Comment (maximecb) [inline yjit/src/codegen.rs:1351]: I would have expected the type in the context to be refined to `Object` after this type check, it feels like that should be part of this guard operation so that it happens automatically.

Maybe that should be done as part of a separate PR because I realize it's orthogonal to the change you're making here, but we do that in `guard_two_fixnums`, but we don't do it in `guard_object_is_array` for instance, which means we're not taking full advantage of the checks we're doing.

Comment (tenderlove) [reply yjit/src/codegen.rs:1351]: TBH I didn't really want to write this guard 😅

If we integrate T_OBJECT in to the shape, then we don't need to have this guard.  I decided to see how hard it would be to add T_OBJECT and [it wasn't too bad](https://github.com/ruby/ruby/pull/6758).

Once I merge that, I'll just delete this code.

**Review by maximecb — APPROVED** [review]

File: shape.c
Comment (tenderlove) [reply shape.c]: Renamed!

**Comment by tenderlove** [comment]
I've updated this against latest master.  The new machine code is like this:

```
== BLOCK 1/4, ISEQ RANGE [0,3), 41 bytes ======================
  # getinstancevariable
  0x5560c986d023: mov rax, qword ptr [r13 + 0x18]
  # guard object is heap
  0x5560c986d027: test al, 7
  0x5560c986d02a: jne 0x5560c986f02d
  0x5560c986d030: cmp rax, 4
  0x5560c986d034: jbe 0x5560c986f02d
  # guard shape
  0x5560c986d03a: cmp word ptr [rax + 6], 0x19
  0x5560c986d03f: jne 0x5560c986f046
  0x5560c986d045: mov rax, qword ptr [rax + 0x10]
  0x5560c986d049: mov qword ptr [rbx], rax

== BLOCK 2/4, ISEQ RANGE [3,6), 0 bytes =======================
== BLOCK 3/4, ISEQ RANGE [3,6), 23 bytes ======================
  # gen_direct_jmp: fallthrough
  # getinstancevariable
  # regenerate_branch
  # getinstancevariable
  # regenerate_branch
  0x5560c986d04c: mov rax, qword ptr [r13 + 0x18]
  # guard shape
  0x5560c986d050: cmp word ptr [rax + 6], 0x19
  0x5560c986d055: jne 0x5560c986f099
  0x5560c986d05b: mov rax, qword ptr [rax + 0x18]
  0x5560c986d05f: mov qword ptr [rbx + 8], rax
```

Benchmarks are like this:

```
before: ruby 3.2.0dev (2022-11-18T18:03:33Z master ba3b40a9ae) +YJIT [x86_64-linux]
after: ruby 3.2.0dev (2022-11-18T18:28:20Z 32-bit-compare 8190d35c5d) +YJIT [x86_64-linux]

-------------  -----------  ----------  ----------  ----------  ------------  -------------
bench          before (ms)  stddev (%)  after (ms)  stddev (%)  before/after  after 1st itr
30k_ifelse     230.2        0.0         231.1       1.0         1.00          1.00         
30k_methods    560.8        0.0         558.6       0.0         1.00          1.00         
activerecord   69.7         0.2         69.2        0.2         1.01          1.00         
binarytrees    135.9        0.3         139.7       1.6         0.97          0.98         
cfunc_itself   30.5         0.2         30.5        0.1         1.00          1.00         
chunky_png     461.2        0.3         431.8       0.4         1.07          1.06         
erubi          197.7        0.7         197.7       0.9         1.00          1.00         
erubi_rails    12.1         4.0         12.1        3.4         1.00          1.00         
etanni         330.6        0.4         331.8       0.3         1.00          1.00         
fannkuchredux  1394.8       0.2         1400.6      0.6         1.00          1.04         
fib            41.6         0.9         41.5        0.1         1.00          1.01         
getivar        23.3         0.1         24.5        0.1         0.95          1.00         
hexapdf        1457.7       1.5         1460.8      2.1         1.00          1.00         
keyword_args   38.1         0.2         38.2        0.1         1.00          1.00         
lee            624.4        0.8         633.7       1.0         0.99          1.00         
liquid-render  76.1         1.8         76.1        1.6         1.00          1.00         
mail           105.3        0.3         105.5       0.1         1.00          1.00         
nbody          59.4         1.4         58.8        0.4         1.01          1.02         
optcarrot      2155.9       0.6         2018.3      0.6         1.07          1.07         
psych-load     1358.8       0.1         1354.6      0.1         1.00          1.00         
railsbench     1279.3       1.1         1273.8      2.1         1.00          1.00         
respond_to     21.6         0.2         22.0        0.1         0.98          0.99         
ruby-lsp       50.6         19.5        50.3        20.3        1.01          1.00         
rubykon        4756.7       0.4         4832.8      0.4         0.98          0.98         
setivar        30.0         0.3         30.4        1.0         0.99          1.01         
str_concat     26.6         1.6         26.7        2.5         0.99          1.00         
-------------  -----------  ----------  ----------  ----------  ------------  -------------
Legend:
- before/after: ratio of before/after time. Higher is better for after. Above 1 represents a speedup.
- after 1st itr: ratio of before/after time for the first benchmarking iteration.
```

**Review by maximecb — APPROVED** [review]
Thanks for bearing with me and well done Aaron. The new ivar read code looks very nice 👌

**Comment by tenderlove** [comment]
> Thanks for bearing with me and well done Aaron. The new ivar read code looks very nice. It's about as efficient as it gets👌

🙇🏻‍♀️

I think if we could make the ctx learn the current shape, we could generate one shape guard and amortize that over multiple reads.  So maybe we could get IV reads to two instructions?

**Comment by XrXr** [comment]
Nice! By the way since there is no extra T_OBJECT guard now the commit message is outdated.

**Comment by tenderlove** [comment]
> Nice! By the way since there is no extra T_OBJECT guard now the commit message is outdated.

Thanks. Thought I had removed it everywhere.  I'll fix the commit message after the build comes back green (I don't want to wait for CI again) 🙇

**Comment by maximecb** [comment]
> I think if we could make the ctx learn the current shape, we could generate one shape guard and amortize that over multiple reads. So maybe we could get IV reads to two instructions (one instruction for reading, the other to push on the stack)?

Yes. Though there's a (slight) tradeoff in terms of memory usage (making the context object bigger). Might also make sense to ask how common it is to have multiple ivar reads in a sequence.

We'll need to redesign the context object next year to do inlining, so maybe we could also take that opportunity to think about how we can encode more useful information into it, such as shape IDs (if known) and also constant values (small integers, Qundef, etc.).

**Pull request merged**
